### PR TITLE
[EventDispatcher] Handle laziness internally instead of relying on ClosureProxyArgument

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22970
| License       | MIT
| Doc PR        | -

If we decide to go this way, we might drop ClosureProxyArgument entirely because we won't need it internally. I'll propose it in another PR for discussion if this one is accepted.

This PR allows "high-order" listeners, as `array(Closure, method)`. Closure is called to get the actual instance when needed only.

How does it look to you? (I'll add tests once confirmed)

### CODE REVIEW

**Review by chalasr — APPROVED** [review]
I think its a nice have for the dispatcher itself and that the bit of complexity added here isn't a big deal as it allows removing much more from DI. 👍 
(adding some tests for this  would be nice I guess)

**Review by ogizanagi — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
now with simplified logic

**Comment by Gemineye** [comment]
This patch seems to fix my Issue in #22970.  👍

File: src/Symfony/Component/EventDispatcher/EventDispatcher.php [hunk 1]
```diff
@@ -110,13 +131,30 @@ public function addListener($eventName, $listener, $priority = 0)
...
+                    unset($listeners[$k], $this->sorted[$eventName]);
+                } else {
+                    $listeners[$k] = $v;
+                }
+            }
+
+            if ($listeners) {
+                $this->listeners[$eventName][$priority] = $listeners;
+            } else {
+                unset($this->listeners[$eventName][$priority]);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/EventDispatcher/EventDispatcher.php]: this "unset" allows better cleanup or the `$this->listeners` property and explains the replacement of "isset" by "empty" in a few places above

**Comment by weaverryan** [comment]
This also appears to fix #23011

File: src/Symfony/Component/EventDispatcher/EventDispatcher.php [hunk 2]
```diff
@@ -110,13 +131,30 @@ public function addListener($eventName, $listener, $priority = 0)
...
+            foreach ($listeners as $k => $v) {
+                if ($v !== $listener && is_array($v) && isset($v[0]) && $v[0] instanceof \Closure) {
+                    $v[0] = $v[0]();
+                }
+                if ($v === $listener) {
+                    unset($listeners[$k], $this->sorted[$eventName]);
+                } else {
+                    $listeners[$k] = $v;
+                }
+            }
```
Comment (weaverryan) [inline src/Symfony/Component/EventDispatcher/EventDispatcher.php]: Is this identical to the code down in `sortListeners()`? If so, perhaps a `private function resolveListeners($eventName)` or something similar? I'm finding some of the logic a bit hard to follow - organizing into a private function(s) might also help me understand things :)

**Review by weaverryan — COMMENTED** [review]
The `isset` -> `empty` change added some complexity, at least for reviewing (could that be done in a separate PR?). But, it makes sense to me, looks good and obviously all the existing tests still pass

File: src/Symfony/Component/EventDispatcher/EventDispatcher.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/EventDispatcher/EventDispatcher.php]: yep, the code behavior is the same (this is just a "manual" version of array_merge)
not sure about the private: performance is critical in this components

**Comment by nicolas-grekas** [comment]
> could that be done in a separate PR?

not without comflicts, which means process overhead, I'd prefer not

**Comment by weaverryan** [comment]
👍

File: src/Symfony/Component/EventDispatcher/EventDispatcher.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/EventDispatcher/EventDispatcher.php]: oh, and this most importantly helps the new implem of hasListeners (which is lazy-compliant now)

**Comment by dmaicher** [comment]
Some tests would be good for it I think :smile: :+1:

File: src/Symfony/Component/EventDispatcher/EventDispatcher.php [hunk 3]
```diff
@@ -110,13 +131,30 @@ public function addListener($eventName, $listener, $priority = 0)
...
-            if (false !== ($key = array_search($listener, $listeners, true))) {
-                unset($this->listeners[$eventName][$priority][$key], $this->sorted[$eventName]);
+            foreach ($listeners as $k => $v) {
+                if ($v !== $listener && is_array($v) && isset($v[0]) && $v[0] instanceof \Closure) {
+                    $v[0] = $v[0]();
+                }
+                if ($v === $listener) {
+                    unset($listeners[$k], $this->sorted[$eventName]);
+                } else {
+                    $listeners[$k] = $v;
```
Comment (dmaicher) [inline src/Symfony/Component/EventDispatcher/EventDispatcher.php]: can we not use `foreach ($listeners as $k => &$v) {` instead?

Comment (nicolas-grekas) [inline src/Symfony/Component/EventDispatcher/EventDispatcher.php]: references don't play well with php 5, I try to avoid them as much as possible

**Review by fabpot — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Now with tests

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by vudaltsov** [comment]
@nicolas-grekas , am I right that now if I create a subscriber like this:

```php
<?php

namespace AppBundle\EventListener;

use Symfony\Component\EventDispatcher\EventSubscriberInterface;

class TestListener implements EventSubscriberInterface
{
    public function __construct()
    {
        dump('constructor');
    }

    public static function getSubscribedEvents()
    {
        return [
            'test' => 'test',
        ];
    }

    public function test()
    {
        dump('fired');
    }
}
```

I should not see `constructor` in the profiler's dump section until the `test` event is fired?

**Comment by stof** [comment]
@vudaltsov not exactly. The profiler itself triggers initialization currently, due to collecting not-called listeners.

**Comment by vudaltsov** [comment]
@stof , thank you! Now I understand what the problem was... With profiler disabled everything works as expected, so I can rely on this in production.
